pallets/price-oracle/src/lib.rs

```
@@ -0,0 +1,231 @@
+	#[pallet::storage]
+	#[pallet::getter(fn price_accumulator)]
+	pub type PriceDataAccumulator<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, PriceEntry, ValueQuery>;
+
+	/// The last ten average values corresponding to the last ten blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_ten)]
+	pub type PriceDataTen<T: Config> = StorageValue<_, Vec<(AssetPairId, BucketQueue)>, ValueQuery>;
+
+	/// The last ten average values corresponding to the last hundred blocks.
+	/// Each average value corresponds to an interval of length ten blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_hundred)]
+	pub type PriceDataHundred<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	/// The last ten average values corresponding to the last thousand blocks.
+	/// Each average value corresponds to an interval of length hundred blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_thousand)]
+	pub type PriceDataThousand<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// update average values in the storage
+			Self::update_data();
+			// clear the price buffer
+			PriceDataAccumulator::<T>::remove_all(None);
+
+			T::WeightInfo::on_initialize_one_entry_multiple_tokens(Self::num_of_assets())
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn on_create_pool(asset_pair: AssetPair) {
+		let data = PriceDataTen::<T>::get();
+		if !data.iter().any(|bucket_tuple| bucket_tuple.0 == asset_pair.name()) {
+			let incremented_asset_count = if let Some(count) = Self::num_of_assets().checked_add(1) {
+				count
+			} else {
+				// We don't want to throw an error here because this method is used in different extrinsics.
+				// We also do not expect to have more than 2^32 assets registered.
+				return
```

I don't think that just returning here is correct. I believe you can panic here. 

Because if you reach this state - although unlikely - the stuff will not work properly anymore and you'd fail silently.